File: 1. 단계별로_풀어보기/30~39/33. 트리/7. 4803.py
import sys

sys.setrecursionlimit(100000)

input = sys.stdin.readline
output = sys.stdout.write

parentArr = []
connInfoArr = []

def isTree(parent, now):
  result = True
  for child in connInfoArr[now]:
    if child != parent:
      if parentArr[child] != 0:
        result = False
      else:
        parentArr[child] = now
        temp = isTree(now, child)
        result = result & temp

  return result

caseIdx = 1
while True:
  N, M = map(int, input().split())

  if N == 0:
    break

  connInfoArr = list([] for i in range(N + 1))
  parentArr = list(0 for i in range(N + 1))

  for m in range(M):
    A, B = map(int, input().split())
    connInfoArr[A].append(B)
    connInfoArr[B].append(A)

  numTree = 0

  for i in range(1, N + 1):
    if parentArr[i] == 0:
      parentArr[i] = i
      numTree += 1 if isTree(i, i) else 0

  if numTree > 1:
    output(f"Case {caseIdx}: A forest of {numTree} trees.\n")
  elif numTree == 1:
    output(f"Case {caseIdx}: There is one tree.\n")
  else:
    output(f"Case {caseIdx}: No trees.\n")

  caseIdx += 1
  

File: 1. 단계별로_풀어보기/33. 트리/7. 4803.js
// 입력
const fs = require('fs')
const inputArr = fs.readFileSync(0).toString().trim().split('\n')

// 글로벌 변수처럼 쓸 변수들
let treeInfoArr = []
let connInfoArr = []
let parentArr = new Array(501)

// 입력 파싱 및 연산 수행
let remain = 0
let cntCase = 1
inputArr.map((inputRow, rowIdx) => {
  if (remain === 0) {
    const [N, M] = inputArr[rowIdx].trim().split(' ').map(Number)

    if (N === 0) {
      return null
    }

    treeInfoArr = inputArr.slice(rowIdx + 1, rowIdx + M + 1).map(row => row.trim().split(' ').map(Number))
    connInfoArr = Array.from({length: N + 1}, () => [])
    parentArr.fill(0)

    treeInfoArr.forEach(info => {
      const [A, B] = info
      connInfoArr[A].push(B)
      connInfoArr[B].push(A)
    })

    printQueryResult(cntCase, N)
    cntCase++

    remain = M
  } // ::
  else {
    remain--
    return null
  }
})

function printQueryResult(cntCase, N) {
  let numTree = 0
  for (let i = 1; i <= N; i++) {
    if (parentArr[i] === 0) {
      parentArr[i] = i
      numTree += isTree(i, i) ? 1 : 0
    }
  }

  if (numTree > 1) {
    process.stdout.write(`Case ${cntCase}: A forest of ${numTree} trees.\n`)
  } // ::
  else if (numTree === 1) {
    process.stdout.write(`Case ${cntCase}: There is one tree.\n`)
  } // ::
  else {
    process.stdout.write(`Case ${cntCase}: No trees.\n`)
  }
}

function isTree(parent, now) {
  let result = true

  for (let i = 0; i < connInfoArr[now].length; i++) {
    const child = connInfoArr[now][i]

    if (child !== parent) {
      if (parentArr[child] !== 0) {
        result = false
      } // ::
      else {
        parentArr[child] = now
        const tempResult = isTree(now, child)
        result &= tempResult
      }
    }
  }

  return result
}
